In [1]:
import pandas as pd
import seaborn as sns
import numpy as np

In [2]:
df=pd.read_csv("/content/drive/MyDrive/internity internship/Machine learning/StudentsPerformance.csv")
df.head()

gender race/ethnicity  ... reading score writing score
0  female        group B  ...            72            74
1  female        group C  ...            90            88
2  female        group B  ...            95            93
3    male        group A  ...            57            44
4    male        group C  ...            78            75

[5 rows x 8 columns]

In [3]:
df.rename(columns={"parental level of education": "parental_education","test preparation course":"preparation_status","math score":"maths","reading score":"reading","writing score":"writing"},inplace=True)
df.describe()

maths      reading      writing
count  1000.00000  1000.000000  1000.000000
mean     66.08900    69.169000    68.054000
std      15.16308    14.600192    15.195657
min       0.00000    17.000000    10.000000
25%      57.00000    59.000000    57.750000
50%      66.00000    70.000000    69.000000
75%      77.00000    79.000000    79.000000
max     100.00000   100.000000   100.000000

In [4]:
df.head()

gender race/ethnicity  parental_education  ... maths reading  writing
0  female        group B   bachelor's degree  ...    72      72       74
1  female        group C        some college  ...    69      90       88
2  female        group B     master's degree  ...    90      95       93
3    male        group A  associate's degree  ...    47      57       44
4    male        group C        some college  ...    76      78       75

[5 rows x 8 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   gender              1000 non-null   object
 1   race/ethnicity      1000 non-null   object
 2   parental_education  1000 non-null   object
 3   lunch               1000 non-null   object
 4   preparation_status  1000 non-null   object
 5   maths               1000 non-null   int64 
 6   reading             1000 non-null   int64 
 7   writing             1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [6]:
df.nunique()

gender                 2
race/ethnicity         5
parental_education     6
lunch                  2
preparation_status     2
maths                 81
reading               72
writing               77
dtype: int64

NO OF BOYS AND GIRL APPEARED FOR EXAM

In [7]:
df.groupby("gender")["gender"].count()

gender
female    518
male      482
Name: gender, dtype: int64

**PERFORMANCE OF STUDENT IN EXAM WITH RESPECT TO GENDER**        

*DEFINED A FUNTION THAT GIVE THE OUTPUT WHO PERFORMED BETTER*

In [8]:
df_1=df.copy()
boys=(df_1[(df_1["gender"]=="male")&
     (df_1["maths"]>=df.maths.quantile(0.75))&
     (df_1["reading"]>=df.reading.quantile(0.75))&
     (df_1["writing"]>=df.writing.quantile(0.75))
     ]).gender.count()

In [9]:
df_1=df.copy()
girls=(df_1[(df_1["gender"]=="female")&
     (df_1["maths"]>=df.maths.quantile(0.75))&
     (df_1["reading"]>=df.reading.quantile(0.75))&
     (df_1["writing"]>=df.writing.quantile(0.75))
     ]).gender.count()

In [10]:
def performance(girls,boys):
     if girls>boys :
         return ("Girls perform well. ")
     else :
         return ("Boys played well. ")    

In [11]:
performance(girls,boys)

'Girls perform well. '

**# NO OF STUDENT PASSED ACCORDING TO GENDER**

In [12]:
df[(df["maths"]>=60)&
   (df["reading"]>=60)&
   (df["writing"]>=60)].groupby("gender")["gender"].count()

gender
female    319
male      284
Name: gender, dtype: int64

In [13]:
df.head()

gender race/ethnicity  parental_education  ... maths reading  writing
0  female        group B   bachelor's degree  ...    72      72       74
1  female        group C        some college  ...    69      90       88
2  female        group B     master's degree  ...    90      95       93
3    male        group A  associate's degree  ...    47      57       44
4    male        group C        some college  ...    76      78       75

[5 rows x 8 columns]

In [14]:
def percent(a,b,c):
    return((a+b+c)/3)

In [15]:
df["percentage"]=(df["maths"]+df["reading"]+df["writing"])/3

In [16]:
df

gender race/ethnicity  parental_education  ... reading writing  percentage
0    female        group B   bachelor's degree  ...      72      74   72.666667
1    female        group C        some college  ...      90      88   82.333333
2    female        group B     master's degree  ...      95      93   92.666667
3      male        group A  associate's degree  ...      57      44   49.333333
4      male        group C        some college  ...      78      75   76.333333
..      ...            ...                 ...  ...     ...     ...         ...
995  female        group E     master's degree  ...      99      95   94.000000
996    male        group C         high school  ...      55      55   57.333333
997  female        group C         high school  ...      71      65   65.000000
998  female        group D        some college  ...      78      77   74.333333
999  female        group D        some college  ...      86      86   83.000000

[1000 rows x 9 columns]

In [17]:
X=df.iloc[:,5:8].values
Y=df.iloc[:,8].values

In [18]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.30)

In [ ]:
print(x_train.shape)
print(y_train.shape)

(700, 3)
(700,)


In [ ]:
print(x_test.shape)
print(y_test.shape)

(300, 3)
(300,)


In [21]:
from sklearn import linear_model
regr = linear_model.LinearRegression()
regr.fit(X, Y)
print(regr.coef_)

[0.33333333 0.33333333 0.33333333]


The result array represents the coefficient values of maths , reading, writing marks.

Maths: 0.3333333
Reading: 0.3333333
Writing: 0.3333333
These values tell us that if the maths marks increase by 1, the percentage  increases by 0.33333333.

And if the reading marks increase by 1, the percentage  increases by 0.33333333.

In [ ]:
n=len(X)

In [ ]:
mean_x=np.mean(X)
mean_y=np.mean(Y)

In [ ]:
num=0
denom=0
for i in range(n):
    num +=(X[i]-mean_x)*(Y[i]-mean_y)
    denom +=(X[i]-mean_x)**2
m=num/denom
c=mean_y-(m*mean_x)

In [ ]:
print(m,c)

[0.85335842 0.93892256 0.90572214] [9.93799731 4.13925915 6.38927368]


In [ ]:
m

array([0.85335842, 0.93892256, 0.90572214])